### DISCUSSION TITLE
UTF-8 and substring(string, start, length) - bytes or characters ?

### DISCUSSION THREAD

**Discussion by jimdickson007** [duckdb/duckdb#14169] [Q&A] [answered]
I need to read a fixed-width file that contains UTF-8 characters.
I plan to use substring to parse each record according to a DDL of offsets.

Qu: Do the  start & length arguments refer offsets in bytes or characters ?

For reference, Oracle offers substr, substrB, substrC (Bytes/Characters)

[https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SUBSTR.html](url)

PS I know UTF-8 allows multi-byte characters so believe this path leads to madness but a bank is involved :(

**Comment by Tishj** [comment, marked as answer]
I feel like you're looking for `substring_grapheme`

This offers some additional context:
<https://github.com/duckdb/duckdb/pull/5028>

**Reply by jimdickson007** [reply]
Excellent - I did see that, but did not understand implications ( think docs need clarification ).
The context you posted helps.

Suspect this is the answer, but let me try it out first, will update later.
( Need data provider to specify whether offsets they have supplied are bytes or characters plus lots of testing with a Thai reader )

**Reply by Tishj** [reply]
Cool, yea maybe you can make a duckdb-web issue, I also don't think this is properly documented. I just remembered there was a decently sized PR related to utf8 length/indexing functions

**Reply by jimdickson007** [reply]
Haha - I have no idea what "duckdb-web issue" means - I might be too old skool.
Anyway will mark this as answered even though I have not had time to test yet.
